local Types = require("@ci/test-types.luau")

return function(test: Types.Test, expect: Types.Expect)
    test("new", function()
        local new = GameObject.new("Mesh")
        expect(new:HasComponent("Mesh")):toBe(true)
        expect(new:HasComponent("RigidBody")):toBe(true)
        expect(new:HasComponent("Transform")):toBe(true)
        expect(new.Name):toBe("Mesh")
        new:Destroy()
    end)
    test("fromComponents", function()
		local new = GameObject.fromComponents({ "SpotLight", "AnimationAsset", "Transform" })
		expect(new:HasComponent("SpotLight")):toBe(true)
        expect(new:HasComponent("Transform")):toBe(true)
        expect(new:HasComponent("AnimationAsset")):toBe(true)
		expect(new.Name):toBe("GameObject")
		new:Destroy()
	end)
    test("fromId", function()
        expect(GameObject.fromId(game.ObjectId)):toBe(game)
    end)
    test("validComponents", function()
        expect(#GameObject.validComponents > 5):toBe(true)
        for i, v in GameObject.validComponents do
            expect(i > 0):toBe(true)
            expect(typeof(v) == "string")
        end
    end)
    test("new - no args", function()
        local new = GameObject.new()
        expect(#new:GetComponents()):toBe(0)
        new:Destroy()
    end)
    test("fromComponents - no args", function()
        local new = GameObject.fromComponents()
        expect(#new:GetComponents()):toBe(0)
        new:Destroy()
    end)
    test("fromComponents - empty table", function()
        local new = GameObject.fromComponents({})
        expect(#new:GetComponents()):toBe(0)
        new:Destroy()
    end)
end
